chalk-recursive/src/recursive.rs

```diff
@@ -224,20 +239,40 @@ impl<'me, I: Interner> SolveDatabase<I> for Solver<'me, I> {
             // Check if this table is still on the stack.
             if let Some(depth) = self.context.search_graph[dfn].stack_depth {
                 // Is this a coinductive goal? If so, that is success,
-                // so we can return normally. Note that this return is
-                // not tabled.
-                //
-                // XXX how does caching with coinduction work?
+                // so we can return and set the minimum to its DFN.
+                // Note that this return is not tabled. And so are
+                // all other solutions in the cycle until the cycle
+                // start is finished. This avoids prematurely cached
+                // false positives.
                 if self.context.stack.coinductive_cycle_from(depth) {
```

Comment:
I think this should be a bit different. I think ideally it would be like:

```rust
 if let Some(depth) = self.context.search_graph[dfn].stack_depth {
    self.context.stack[depth].flag_cycle();
    if self.context.stack.mixed_inductive_coinductive_cycle_from(depth) {
        return Error();
    }
}
```

(I don't know whether flagging the cycle is needed or not in the case of error, but I don't see how it can hurt)

Reply:
note that the coinductive case (like the inductive case) just falls through